### PR TITLE
Limit to endpoint feature

### PR DESCRIPTION
**Pull request by ealsur**
# Pull Request Template

## Description

With `CosmosClientOptions.LimitToEndpoint`, the CosmosClient will limit all operations to the endpoint specified on the CosmosClient constructor. If that endpoint is a regional endpoint, it means that the account behaves as a single region account for the purposes of availability.

`CosmosClientBuilder.WithLimitToEndpoint(bool)` has the same effect as part of the CosmosClientBuilder.

This feature is not compatible with ApplicationRegion, if one is set, the other can't be used and it will be validated through an `ArgumentException`.

## Type of change

- [X] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs [hunk 1]
```diff
@@ -349,6 +349,18 @@ public CosmosSerializer Serializer
...
 
+        /// <summary>
+        /// Limits the operations to the provided endpoint on the CosmosClient.
+        /// </summary>
+        /// <value>
+        /// Default value is false.
+        /// </value>
+        /// <remarks>
+        /// When the value of this property is false, the SDK will automatically discover write and read regions, and use them when the configured application region is not available. When set to true, availability is limited to the endpoint specified on the CosmosClient constructor and defining the application region is not allowed.
+        /// </remarks>
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: New line for "When true..."

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs [hunk 2]
```diff
@@ -349,6 +349,18 @@ public CosmosSerializer Serializer
...
         }
 
+        /// <summary>
+        /// Limits the operations to the provided endpoint on the CosmosClient.
+        /// </summary>
+        /// <value>
+        /// Default value is false.
+        /// </value>
+        /// <remarks>
+        /// When the value of this property is false, the SDK will automatically discover write and read regions, and use them when the configured application region is not available. When set to true, availability is limited to the endpoint specified on the CosmosClient constructor and defining the application region is not allowed.
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: >and defining the application region is not allowed. [](start = 284, length = 51)

New statement for this please.

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs [hunk 3]
```diff
@@ -349,6 +349,18 @@ public CosmosSerializer Serializer
...
+        /// </remarks>
+        /// <seealso href="https://docs.microsoft.com/azure/cosmos-db/high-availability"/>
+        public bool LimitToEndpoint { get; set; }
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: How about a example usage.

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs [hunk 4]
```diff
@@ -498,6 +510,7 @@ internal CosmosClientOptions Clone()
         internal ConnectionPolicy GetConnectionPolicy()
         {
             this.ValidateDirectTCPSettings();
+            this.ValidateLimitToEndpointSettings();
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: Thinking loud. Thoughts on any patterns which can validate the group applicability? 
Not a blocker for this release.

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs (hunk 3 above)
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: Please state the default 'false' explictly

Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: UT is in place in the PR already, will add default value

Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: Example usage is on the Builder

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/CosmosClientOptionsUnitTests.cs [hunk 5]
```diff
@@ -68,6 +68,7 @@ public void VerifyCosmosConfigurationPropertiesGetUpdated()
             Assert.IsNull(clientOptions.SerializerOptions);
             Assert.IsNull(clientOptions.Serializer);
             Assert.IsNull(clientOptions.WebProxy);
+            Assert.IsFalse(clientOptions.LimitToEndpoint);
 
             //Verify GetConnectionPolicy returns the correct values for default
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/CosmosClientOptionsUnitTests.cs]: Please get sign-off from Ji.

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs (hunk 2 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: Done

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs (hunk 1 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: Done

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs (hunk 4 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: Well, we could implement some sort of validate attribute, but the validation would need to take place during build time. And the validation attribute would need to describe which is the check that needs to be performed on the conflicting attributes?

**Pull request merged**